File: clients/unity-app/Assets/Scripts/Labs/Lab03.cs
using System.Threading.Tasks;
using Colyseus;
using Colyseus.Predict;
using ColyseusSchema = Colyseus.Schema.Schema;
using UnityEngine;
using Lab = PredictProbe.LabSchema;

namespace Playground
{
    /// <summary>
    /// Lab 03 — Predict &amp; Reconcile. The reconciler OBSERVES the input handle:
    /// every Send() is predicted locally the same instant and buffered. When the
    /// server's next patch acks input N, it rewinds to the authoritative state
    /// and REPLAYS inputs N+1.. through the same shared StepEntity — so the
    /// predicted pose stays consistent with everything the server hasn't seen.
    /// Port of labs/03-reconcile/.
    /// </summary>
    public class Lab03 : LabBase<Lab.MoveState>
    {
        public override string Id => "03-reconcile";
        public override int Num => 3;
        public override string Title => "Predict & Reconcile";
        public override string Blurb => "Rollback to the ack, replay pending inputs, smooth the error.";

        private readonly MoveLane _lane = new MoveLane();
        private bool _renderSmoothed = true, _showGhost = true;
        private readonly Spark _driftSpark = new Spark();
        private double _sparkGate;


        /// <summary>The lane, for the acceptance harness to read telemetry off.</summary>
        public MoveLane Lane => _lane;

        public override async Task<bool> Mount(App app)
        {
            if (!await _lane.Join(app)) return false;
            Room = _lane.Room;
            return true;
        }

        public override void Frame(App app, double now, double dtMs)
        {
            if (Kb.Key(KeyCode.I)) Room.Send("impulse");
            if (Kb.Key(KeyCode.T)) Room.Send("teleport");
            if (Kb.Key(KeyCode.V)) _renderSmoothed = !_renderSmoothed;
            if (Kb.Key(KeyCode.G)) _showGhost = !_showGhost;
            if (Kb.Key(KeyCode.N)) _lane.AutoSnap = !_lane.AutoSnap;
            int smoothStep = Kb.Key(KeyCode.Equals) ? 10 : Kb.Key(KeyCode.Minus) ? -10 : 0;
            if (smoothStep != 0)
                _lane.Build(Mathf.Clamp((float)(_lane.SmoothMs + smoothStep), 0, 200));

            _lane.Drive(now, Kb.MoveX(), Kb.MoveY());

            _sparkGate += dtMs;
            if (_sparkGate >= 100) { _sparkGate = 0; _driftSpark.Push(_lane.Recon.Drift.Ema); }
        }

        public override void Render(App app)
        {
            var v = app.View;
            var me = _lane.Me;
            Room.State.players.ForEach((key, p) =>
            {
                if (key == _lane.Sid) return;
                Draw.Square(v, _lane.RemoteX(p), _lane.RemoteY(p), Sim.PlayerHalf, Palette.Hue(p.hue, 0.45f));
            });

            // Server ghost: the raw authoritative pose — it trails by ~RTT.
            if (_showGhost)
            {
                Draw.GhostSquare(v, me.x, me.y, Sim.PlayerHalf, Palette.A(Palette.Text, 0.75f));
                Draw.Label(v, me.x, me.y, "server", Palette.A(Palette.Text, 0.55f), 11,
                    v.S(Sim.PlayerHalf) + 4);
            }

            double px = _renderSmoothed ? _lane.X : _lane.Recon.State.x;
            double py = _renderSmoothed ? _lane.Y : _lane.Recon.State.y;
            Draw.Square(v, px, py, Sim.PlayerHalf, Palette.Hue(me.hue));
            Draw.SquareOutline(v, px, py, Sim.PlayerHalf, Palette.Text);
            Draw.Label(v, px, py, "you (predicted)", Palette.Text, 11, -v.S(Sim.PlayerHalf) - 18);

            var recon = _lane.Recon;
            var status = DriftClassifier.Classify(recon.Drift, 0);
            var h = app.Hud;
            h.Section("TELEMETRY");
            h.Chips("pending inputs (unacked)", recon.PendingCount);
            h.Row("drift status", status.ToString().ToLower(),
                status == DriftStatus.Matched ? Palette.Good :
                status == DriftStatus.Jitter ? Palette.Warn : Palette.Bad);
            h.Spark(_driftSpark, "drift ema", $"{recon.Drift.Ema:F4}", Palette.Accent);
            h.Row("last correction", $"{recon.LastCorrectionMag:F3}", Palette.Text);
            h.Row("corrections seen", _lane.Corrections.ToString(), Palette.Text);
            h.Row("reconciles", recon.ReconcileSeq.ToString(), Palette.Text);

            h.Section("CONTROLS");
            h.Key("WASD", "drive");
            h.Key("I", "force mispredict (impulse)");
            h.Key("T", "teleport");
            h.Key("- / =", $"smoothing {_lane.SmoothMs:F0} ms");
            h.Key("V", _renderSmoothed ? "render: Value() smoothed" : "render: State (exact)");
            h.Key("G", _showGhost ? "server ghost: on" : "server ghost: off");
            h.Key("N", _lane.AutoSnap ? "snap on teleport: on" : "snap on teleport: off");
            h.Note("Corrections beyond 8u call Reset() — a cut, not a cross-arena glide. " +
                   "Turn it off with N and teleport to see why.");
        }

        public override void Unmount() => _lane.Dispose();

        public override void OnReconnect() => _lane.Rebind();
    }
}

File: clients/defold-app/playground/labs/lab03.lua
--
-- Lab 03 — Predict & Reconcile.
--
-- The reconciler OBSERVES the input handle: every send() is predicted locally
-- the same instant and buffered. When the server's next patch acks input N, it
-- rewinds to the authoritative state and REPLAYS inputs N+1.. through the same
-- shared step_entity — so the predicted pose stays consistent with everything
-- the server hasn't seen.
--
-- Port of src/client/labs/03-reconcile/.
--
local sim = require 'playground.sim'
local app = require 'playground.app'
local MoveLane = require 'playground.move_lane'
local drift = require 'colyseus.predict.drift'

local kb = app.kb

local Lab = {
  id = "03-reconcile",
  num = 3,
  title = "Predict & Reconcile",
  blurb = "Rollback to the ack, replay pending inputs, smooth the error.",
  room_name = "lab-move",
}
Lab.__index = Lab

function Lab.new()
  return setmetatable({
    lane = MoveLane.new(),
    render_smoothed = true,
    show_ghost = true,
  }, Lab)
end

function Lab:ready(room) return MoveLane.ready(room) end

function Lab:mount(context, room)
  self.room = room
  return self.lane:attach(room)
end

function Lab:frame(context, now, dt_ms)
  local lane = self.lane
  if kb.key("i") then self.room:send("impulse") end
  if kb.key("t") then self.room:send("teleport") end
  if kb.key("v") then self.render_smoothed = not self.render_smoothed end
  if kb.key("g") then self.show_ghost = not self.show_ghost end
  if kb.key("n") then lane.auto_snap = not lane.auto_snap end
  local step = kb.key("equals") and 10 or kb.key("minus") and -10 or 0
  if step ~= 0 then
    lane:build(math.max(0, math.min(200, lane.smooth_ms + step)))
  end

  lane:drive(now, kb.move_x(), kb.move_y())
end

function Lab:render(gfx)
  local lane = self.lane
  local me = lane.me

  self.room.state.players:each(function(p, key)
    if key == lane.sid then return end
    gfx.square(lane:remote_x(p), lane:remote_y(p), sim.PLAYER_HALF, gfx.hue(p.hue, 0.45))
  end)

  -- Server ghost: the raw authoritative pose — it trails by ~RTT.
  if self.show_ghost then
    gfx.ghost_square(me.x, me.y, sim.PLAYER_HALF, gfx.a(gfx.PALETTE.text, 0.75))
    gfx.label(me.x, me.y, "server", gfx.a(gfx.PALETTE.text, 0.55), 11, 14)
  end

  local px = self.render_smoothed and lane:x() or lane.recon.state.x
  local py = self.render_smoothed and lane:y() or lane.recon.state.y
  gfx.square(px, py, sim.PLAYER_HALF, gfx.hue(me.hue))
  gfx.square_outline(px, py, sim.PLAYER_HALF, gfx.PALETTE.text, 1)
  gfx.label(px, py, "you (predicted)", gfx.PALETTE.text, 11, -18)

  local recon = lane.recon
  local status = drift.classify(recon.drift, 0)
  gfx.hud_section("TELEMETRY")
  gfx.hud_chips("pending inputs (unacked)", recon:pending_count())
  gfx.hud_row("drift status", status,
    status == "matched" and gfx.PALETTE.good
      or status == "jitter" and gfx.PALETTE.warn
      or gfx.PALETTE.bad)
  gfx.hud_row("drift ema", string.format("%.4f", recon.drift.ema), gfx.PALETTE.accent)
  gfx.hud_row("last correction", string.format("%.3f", recon.last_correction_mag), gfx.PALETTE.text)
  gfx.hud_row("corrections seen", tostring(lane.corrections), gfx.PALETTE.text)
  gfx.hud_row("reconciles", tostring(recon.reconcile_seq), gfx.PALETTE.text)

  gfx.hud_section("CONTROLS")
  gfx.hud_key("WASD", "drive")
  gfx.hud_key("I", "force mispredict (impulse)")
  gfx.hud_key("T", "teleport")
  gfx.hud_key("- / =", string.format("smoothing %.0f ms", lane.smooth_ms))
  gfx.hud_key("V", self.render_smoothed and "render: value() smoothed" or "render: state (exact)")
  gfx.hud_key("G", self.show_ghost and "server ghost: on" or "server ghost: off")
  gfx.hud_key("N", lane.auto_snap and "snap on teleport: on" or "snap on teleport: off")
  gfx.hud_note("Corrections beyond 8u call reset() — a cut, not a cross-arena " ..
    "glide. Turn it off with N and teleport to see why.")
end

function Lab:unmount() self.lane:dispose() end
function Lab:on_reconnect() self.lane:rebind() end

return Lab
